test/e2e/e2e.go

```diff
@@ -102,6 +102,11 @@ var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
 	// unschedulable, we need to wait until all of them are schedulable.
 	framework.ExpectNoError(framework.WaitForAllNodesSchedulable(c, framework.TestContext.NodeSchedulableTimeout))
 
+	// If NumNodes is not specified then auto-detect how many are scheduleable and not tainted
+	if framework.TestContext.CloudConfig.NumNodes == framework.DefaultNumNodes {
+		framework.TestContext.CloudConfig.NumNodes = len(framework.GetReadySchedulableNodesOrDie(c).Items)
```

Comment:
this greatly simplifies things, given a client set exists here.

just to confirm:

the flags values are fetched earlier than this function, here under `init()`:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/e2e_test.go#L56

so at this point the value is defaulted to -1 already?

Reply:
Yes at this point the value of the flag is passed through. So if not specified it's `-1`. If specified then that value is used. Also, if it's specified to be `-1` then this logic is still called.

Reply:
Much better imo.